### PR TITLE
Run `cargo nextest` in addition to `cargo test` in CI

### LINKED ISSUE #787: Try cargo-nextest in CI for better failure reporting
As discussed in [chat](https://matrix.to/#/!uMMSVUNAgMdoTtyJPR:oxide.computer/$syJCPyZqK1kSrZuVyOGvlRYKkwmNNXCY3q-wiuYVh3M?via=oxide.computer&via=unix.house&via=timecube.club), our test output in CI makes it hard to figure out what tests failed. In one very typical case, the failure was buried 500 lines above the bottom of the output. [Running with `--verbose` removed](https://github.com/oxidecomputer/omicron/pull/750) did not help, as there is still a lot of noise. `cargo-nextest` has _much_ nicer output (see below).

Testing this may be as simple as adding a `cargo install cargo-nextest` step to the CI config and updating the test command. Downloading the binary would be faster, but building from source isn't that slow to start out with.

---
 
> Clean, beautiful user interface. nextest presents its results concisely so you can see which tests passed and failed at a glance.

https://nexte.st/index.html

image

image

### PR DESCRIPTION
**Pull request by david-crespo**
[cargo-nextest](https://nexte.st/) is a new Rust test runner. Imagine a world where you can actually tell what tests failed by looking at the CI output. That world is... this world.

As discussed in the [comment](https://github.com/oxidecomputer/omicron/pull/788#issuecomment-1101706057) below, we want to compare nextest and `cargo test` results for a while without committing to making `cargo nextest` results authoritative, so this adds a job that is _not_ required to pass for PRs to get merged. Most of the time the two runners should have the same result, but we want to see if that's really true.

Closes #787.

### Advantages for CI

- `failure-output = "immediate-final"` causes failures to be printed when they happen _and_ again at the end so you can see what failed without scrolling
- JUnit output, which can be archived (done here) and analyzed over time (not done here) and also picked up by another GH action that displays it nicely on the CI results page (not done here)

In theory it should run the test suite [a bit faster](https://nexte.st/book/benchmarks.html) too, but I haven't be able to tell whether we're getting a speedup because the times vary so much.

### Risks

- The library is on the new side and we don't have a lot of experience with it
- Most people are not using it in local dev, so it's possible we could see surprising behavior that fails to repro with `cargo test`

My sense is these are small risks and the project is pretty reliable and stable, but I would want input from people who have a better feel for the ecosystem and what people are using. I believe some of us are using it locally already.

### What this PR does

Add a job `build-and-nextest` that does the following:

- Use a prebuilt action to install `cargo-nextest`
- Use `cargo-nextest` as the test runner for the CI test job
- Have `cargo-nextest` produce JUnit output
  - This requires a config file because this option is not in the CLI
- Archive said JUnit output whether tests pass or fail

### Example failure output

Note that the failure is at the bottom of the output.

image

### CODE REVIEW

**Comment by david-crespo** [comment]
Ok `cargo install nextest` took 3 minutes so that's clearly not tenable.

**Comment by david-crespo** [comment]
Output is an improvement — the failure is only 60 lines up from the bottom. I have a feeling there's a flag we can pass to make it not print these PASSes at the bottom.

image

**Comment by david-crespo** [comment]
Tried with `--no-capture` and it was way worse, so I took that commit back off and force pushed.

<details>
<summary><code>--no-capture</code> output</summary>
image
</details>

**Comment by sunshowers** [comment]
> Output is an improvement — the failure is only 60 lines up from the bottom.

Hi there, author of nextest here! I think your problem can be solved with nextest by passing in `--failure-output immediate-final`: see https://nexte.st/book/other-options.html#reporter-options. You can also define a CI profile in config with this setting: https://nexte.st/book/configuration.html#profiles

**Comment by david-crespo** [comment]
@sunshowers thank you! A bunch of us have already been using nextest locally and love it.

**Comment by david-crespo** [comment]
Using the GH action to get the nextest binary takes 0s. So that's good.

**Comment by david-crespo** [comment]
Output looks good! I'm going to try it with some other failures sprinkled around.

image

File: .cargo/nextest.toml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+[profile.ci]
+fail-fast = false
+failure-output = "immediate-final"
+
+[profile.ci.junit]
+path = "junit.xml"
```
Comment (david-crespo) [inline .cargo/nextest.toml:6]: @jclulow it's very easy to tell it to output JUnit, but what should I do with this artifact? There's already a step where we upload stuff someplace, should we do the same with this?

https://github.com/oxidecomputer/omicron/blob/52dfddb9f5c0fe4e73910b387ffd5cd3e1414a89/.github/workflows/rust.yml#L153-L160

Comment (jclulow) [reply .cargo/nextest.toml:6]: Yes I think we should add a step for uploading the XML file.  I am not especially familiar with this bit, but it seems like we would want it whether the tests fail or not.

```yaml
    - name: Archive test runner JUnit output
      # actions/upload-artifact@v2.3.1
      uses: actions/upload-artifact@82c141cc518b40d92cc801eee768e7aafc9c2fa2
      with:
        name: junit_${{ runner.os }}
        retention-days: 90
        path: |
          junit.xml
```

Something like that maybe?

File: nexus/tests/integration_tests/console_api.rs [hunk 2]
```diff
@@ -100,7 +100,7 @@ async fn test_sessions(cptestctx: &ControlPlaneTestContext) {
         .await
         .expect("failed to get 401 for unauthed API request");
 
-    RequestBuilder::new(&testctx, Method::GET, "/orgs/whatever")
+    RequestBuilder::new(&testctx, Method::GET, "/orgs/nonexistent")
```
Comment (david-crespo) [inline nexus/tests/integration_tests/console_api.rs:103]: forcing tests to run. the run deduper thing is driving me nuts

**Comment by david-crespo** [comment]
Some issues came up in chat. @davepacheco points out that fully changing over to nextest in CI effectively means we are deciding nextest is authoritative, i.e., if `cargo test` and `cargo nextest` give different results in a give case, we listen to `nextest`. Nearly everyone is likely using `cargo test` locally.

For this reason, I'm updating the PR to add separate jobs (well, matrix options on the same job) that run the tests with nextest. This lets us see how often the two test runners give different results, as well as compare output in the case of failures. We can use the generated name of the jobs to keep only `cargo test` required for now.

The required checks would have to be changed in the repo settings to include

```
build-and-test (macos-11, cargo-test)
build-and-test (ubuntu-18.04, cargo-test)
```

image

File: .github/workflows/rust.yml [hunk 3]
```diff
@@ -178,3 +163,78 @@ jobs:
...
+      #   Cargo.lock ensures that everyone is using the same dependencies and
+      #   also gives us a record of which dependencies were used for each CI
+      #   run.  Building with `--locked` ensures that the checked-in Cargo.lock
+      #   is up to date.
+      # - TMPDIR=$OMICRON_TMP: we specify a specific temporary directory so that
+      #   failed test outputs will be in a known place that we can grab at the
+      #   end without also grabbing random other temporary files.
+      run: |
+        TMPDIR=$OMICRON_TMP PATH="$PATH:$PWD/out/cockroachdb/bin:$PWD/out/clickhouse" RUSTFLAGS="-D warnings" RUSTDOCFLAGS="-D warnings" \
+          cargo build --locked --all-targets --verbose
```
Comment (david-crespo) [inline .github/workflows/rust.yml:227]: I think we could skip the `cargo build` step here, but that might make it harder to tell at a glance when the results of these two jobs diverge

**Comment by jclulow** [comment]
Can we add an analogous clone of the buildomat job to run nextest there also?

**Comment by david-crespo** [comment]
Amazingly, it worked on the first try.

**Comment by david-crespo** [comment]
The files in question have changed enough that we'll want to start this from scratch. I still think it would be nice to get dramatically better failure output from CI.

**Pull request closed without merging**